lib/src/ultrahdr_api.cpp

```diff
@@ -0,0 +1,1332 @@
+  }
+}
+
+uhdr_compressed_image_ext::uhdr_compressed_image_ext(uhdr_color_gamut_t cg,
+                                                     uhdr_color_transfer_t ct,
+                                                     uhdr_color_range_t range, unsigned size) {
+  this->m_block = std::make_unique<uhdr_memory_block_t>(size);
+  this->data = this->m_block->m_buffer.get();
+  this->capacity = size;
+  this->data_sz = 0;
+  this->cg = cg;
+  this->ct = ct;
+  this->range = range;
+}
+
+uhdr_error_info_t apply_effects(uhdr_decoder_private* dec) {
+  for (auto& it : dec->m_effects) {
+    std::unique_ptr<ultrahdr::uhdr_raw_image_ext_t> disp_img = nullptr;
+    std::unique_ptr<ultrahdr::uhdr_raw_image_ext_t> gm_img = nullptr;
+
+    if (it.m_type == ultrahdr::uhdr_transform_desc_t::ROTATE) {
+      disp_img = apply_rotate(dec->m_decoded_img_buffer.get(), it.m_attrib.rotate_params.degree);
+      gm_img = apply_rotate(dec->m_gainmap_img_buffer.get(), it.m_attrib.rotate_params.degree);
+    } else if (it.m_type == ultrahdr::uhdr_transform_desc_t::FLIP) {
+      disp_img = apply_flip(dec->m_decoded_img_buffer.get(), it.m_attrib.flip_params.horizontally);
+      gm_img = apply_flip(dec->m_gainmap_img_buffer.get(), it.m_attrib.flip_params.horizontally);
+    } else if (it.m_type == ultrahdr::uhdr_transform_desc_t::CROP) {
+      uhdr_raw_image_t* disp = dec->m_decoded_img_buffer.get();
+      uhdr_raw_image_t* gm = dec->m_gainmap_img_buffer.get();
+      int scale_factor = disp->w / gm->w;
+      int left = it.m_attrib.crop_params.left < 0 ? 0 : it.m_attrib.crop_params.left;
+      int top = it.m_attrib.crop_params.top < 0 ? 0 : it.m_attrib.crop_params.top;
+      int right =
+          it.m_attrib.crop_params.right > (int)disp->w ? disp->w : it.m_attrib.crop_params.right;
+      int bottom =
+          it.m_attrib.crop_params.bottom > (int)disp->h ? disp->h : it.m_attrib.crop_params.right;
+
+      if (right - left <= scale_factor || bottom - top <= scale_factor) {
+        uhdr_error_info_t status;
+        status.error_code = UHDR_CODEC_UNKNOWN_ERROR;
+        status.has_detail = 1;
+        snprintf(status.detail, sizeof status.detail,
+                 "After crop image dimensions are <= 0, display image dimensions %dx%d, gain map "
+                 "image dimensions %dx%d",
+                 right - left, bottom - top, (right - left) / scale_factor,
+                 (bottom - top) / scale_factor);
+        return status;
+      }
+      apply_crop(disp, left, top, right - left, bottom - top);
+      apply_crop(gm, left / scale_factor, top / scale_factor, (right - left) / scale_factor,
+                 (bottom - top) / scale_factor);
+      continue;
+    } else if (it.m_type == ultrahdr::uhdr_transform_desc_t::RESIZE) {
+      uhdr_error_info_t status;
+      status.error_code = UHDR_CODEC_UNSUPPORTED_FEATURE;
+      status.has_detail = 1;
+      snprintf(status.detail, sizeof status.detail, "resize transform is not supported as of now");
```

Comment:
We can have a simple implementation for now, and update it when we bring OpenGL into the library